File: 2000290120141/week1/Linked_List.cpp
#include <iostream>

using namespace std;

struct Node {
  int data;
  Node* next;
};

Node* createNode(int data) {
  Node* newNode = new Node();
  newNode->data = data;
  newNode->next = nullptr;
  return newNode;
}

void insertNode(Node* head, int data) {
  Node* newNode = createNode(data);
  if (head == nullptr) {
    head = newNode;
  } else {
    Node* ptr = head;
    while (ptr->next != nullptr) {
      ptr = ptr->next;
    }
    ptr->next = newNode;
  }
}

void deleteNode(Node* head, int data) {
  if (head == nullptr) {
    return;
  }

  if (head->data == data) {
    Node* temp = head;
    head = head->next;
    delete temp;
    return;
  }

  Node* prev = head;
  Node* curr = head->next;
  while (curr != nullptr && curr->data != data) {
    prev = curr;
    curr = curr->next;
  }

  if (curr == nullptr) {
    return;
  }

  prev->next = curr->next;
  delete curr;
}

void printList(Node* head) {
  Node* ptr = head;
  while (ptr != nullptr) {
    cout << ptr->data << " ";
    ptr = ptr->next;
  }
  cout << endl;
}

int main() {
  Node* head = nullptr;
  insertNode(head, 10);
  insertNode(head, 20);
  insertNode(head, 30);
  printList(head);

  deleteNode(head, 20);
  printList(head);

  return 0;
}

File: 2000290120169/linkedlist.py
#creating linked list 
class node:
    def __init__(self,data):

        self.data=data
        self.next= None 
        
class Linkedlist:
    def __init__(self):
        self.head= None
    
    def listpr(self):
        printval=self.head
        while printval is not None:
            print(printval.data)
            printval=printval.next
    
    def addnode(self,newdata):

        newnode=node(newdata)
        newnode.next=self.head
        self.head=newnode

    def addlast(self,newdata):
        newnode=node(newdata)

        if self.head is None:
            self.head=newnode
            return
        last=self.head 
        while(last.next):
           last= last.next
        last.next=newnode

    def addspecific(self,prev_node,newdata):
        if prev_node is None:
            print('node does not exist')
            return
        newnode=node(newdata)
        newnode.next=prev_node.next
        prev_node.next=newnode

    def deleteNode(self, key):
 
        
        temp = self.head
 
    
        if (temp is not None):
            if (temp.data == key):
                self.head = temp.next
                temp = None
                return
 
        
        
        while(temp is not None):
            if temp.data == key:
                break
            prev = temp
            temp = temp.next
 
    
        if(temp == None):
            return
 
    
        prev.next = temp.next
 
        temp = None
              
l=Linkedlist()
l.head=node(int(input()))
a=node(int(input()))
b=node(int(input()))
c=node(int(input()))
d=node(int(input()))

l.head.next=a
a.next=b
b.next=c
c.next=d 
'''
print(a.next.data)
print(a.data)
print(l.head.data)
print(l.head.next.next.data)
'''
#adding node at beginning
l.addnode(int(input()))
l.addnode(int(input()))


#adding node at last
l.addlast(int(input()))
l.addlast(int(input()))

l.addspecific(a,int(input()))
l.addspecific(b,int(input()))
l.addspecific(c,int(input()))
l.deleteNode(int(input()))

l.listpr()